### ISSUE TITLE
Spurious reference to function literals in 'constant context'

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#3619]
Thanks to @sgrekhov for noticing this!

The definition of what it means to be in a constant context includes a case involving a function literal, [here](https://github.com/dart-lang/language/blob/31c665c48cfc3ea5f0f20fb10b87cbad90840f0b/specification/dartLangSpec.tex#L9090).

However, the situation which is described there can never occur: It is _always_ a compile-time error for a function literal to occur in a constant context, because a function literal is never a constant expression.

This part of the rule is only relevant if we introduce a notion of constant function literals (see https://github.com/dart-lang/language/issues/1048 for a proposal ... from 2012 ... with about 250 supporting emojis).

We may wish to eliminate this part of the 'constant context' rule (or we may wish to add constant function literals). In any case, the current situation is somewhat confusing.

**Comment by lrhn** [comment]
I think it's not *wrong*, and it's deliberately to ensure that a constant context does not cross into a nested function.
The fact that a nested function is a compile-time error for other reasons doesn't change that.

A `const x = [(x) => [x.toString()]];` will be a compile time error because `(x) => [x.toString()]` is not itself a constant expression, but not *also* because `[x.toString()]` occurs in a constant context and is not constant. Because with this clause that's not a constant context. Without the clause, it would be.

I'm not sure being minimal here is desirable.

**Comment by eernstg** [comment]
> I'm not sure being minimal here is desirable.

Makes sense. However, we don't usually specify that a source code construct with an error has any particular properties, we'd usually say/imply that _it isn't relevant_ whether or not `/*1*/` is in a constant context in `const [() { /*1*/ }]`, because `const [() {...}]` is a compile-time error already, no matter what that `...` stands for.

I'd recommend that we change the normative text 'where `e0` is not a function literal (17.11)' to a TODO comment. This ensures that we will have a reminder in case we do add support for constant function literals (we already have another one about `throw` at the same location), and it allows us to avoid the confusing reference to the internals of an expression which is already known to be a compile-time error.

**Comment by lrhn** [comment]
I'd like to introduce the notion of "constant evaluation", as a variant of evaluation that happens in a constant context.
The main reason is to be specific about *where* it happens, and to explicitly make throwing a compile-time error, rather than having the "if evaluation of this expression would throw, it's a compile-time error", which is workable, but a little too indirect. (How do we know if something would throw without evaluating it? If we do evaluate it, why do we say "would throw" instead of "throws"? When are constants evaluated at all, and what does that even *mean*?)

If we do that, an expression in a constant context is evauated using constant evaluation semantics, which is like normal semantics, except that it can end in either a value or a compile-time error. A lot of expression shapes directly evaluate to a compile-time error.

Actually, I think I'd want four modes:
* Constant evaluation
* Required constant evaluation
* Potentially constant evaluation
* Normal evaluation

In a required constant context (optional parameter default values, final field initializer of class with constant constructor), collection literals and object creation expressions are not automatically made constant, like in a constant context.
In a potentially constant context, parameters and type parameters of a surrounding constructor can be accessed too.

Then it's not so much that an expression is in a "constant context", but that the expression is evaluated using *constant evaluation*, which automatically evaluates subexpressions using constant evaluation as well, if the subexpression can be a constant expression. That evaluation will fail when reaching a function literal, and therefore it will never try to evaluate the body in any context. 
We lose the separation between "how must this be evaluated" and "how is this evaluated".

In a normal evaluation context, obvious constants should still be canonicalized (simple literals, type literals with constant type arguments, static/top-level tear-offs, constant instantiations of those).
Which mean we still do recognize *some* inherent constants, as part of normal evaluation, like an extra bit on the static type of an expression: "is constant value", and normal evaluation of `e1 + e2` where `e1` and `e2` evaluate to the *constant* values *v1* and *v2* of type `int`, will evaluate to the constant value *v1*+*v2*. And then we can say that constant `String` values are canonicalized.

That would mean that an expression like `"foo" + "bar"` in a completely normal context could still be canonicalized, because it's adding compile-time-known `String` values.
So both a downwards constant requirement in evaluation and an upwards constant inference/propagation.

(I think it's only really strings where we might *not* canonicalize all possible constant strings today, which is acceptable because *the spec still doesn't explicitly require constant strings to be canonicalized*. It just requires constant `identical` on them to be `true`, but that only applies to string values that can occur in a constant `identical` check.)
